internal/xds/rbac/matchers.go

```diff
@@ -0,0 +1,404 @@
+			if err != nil {
+				return nil, err
+			}
+			matchers = append(matchers, m)
+		case *v3rbacpb.Principal_Header:
+			// Do we need an error here?
+			m, err := newHeaderMatcher(principal.GetHeader())
+			if err != nil {
+				return nil, err
+			}
+			matchers = append(matchers, m)
+		case *v3rbacpb.Principal_UrlPath:
+			m, err := newURLPathMatcher(principal.GetUrlPath())
+			if err != nil {
+				return nil, err
+			}
+			matchers = append(matchers, m)
+		case *v3rbacpb.Principal_NotId:
+			mList, err := matchersFromPrincipals([]*v3rbacpb.Principal{principal})
+			if err != nil {
+				return nil, err
+			}
+			matchers = append(matchers, &notMatcher{
+				matcherToNot: mList[0],
+			})
+		case *v3rbacpb.Principal_SourceIp:
+			// The source ip principal identifier is deprecated. Thus, a
+			// principal typed as a source ip in the identifier will be a no-op.
+			// The config should use DirectRemoteIp instead.
+		case *v3rbacpb.Principal_RemoteIp:
+			// Not supported in gRPC RBAC currently - a principal typed as
+			// Remote Ip in the initial config will be a no-op.
+		case *v3rbacpb.Principal_Metadata:
+			// Not supported in gRPC RBAC currently - a principal typed as
+			// Metadata in the initial config will be a no-op.
+		}
+	}
+	return matchers, nil
+}
+
+// orMatcher is a rbacMatcher where it successfully matches if one of it's children
+// successfully match. It also logically represents a principal or permission,
+// but can also be it's own entity further down the tree of matchers.
+type orMatcher struct {
+	matchers []rbacMatcher
+}
+
+func (om *orMatcher) matches(args *EvaluateArgs) bool {
+	// Range through child matchers and pass in rbacData, and only one child
+	// rbacMatcher has to match to be logically successful.
+	for _, matcher := range om.matchers {
+		if matcher.matches(args) {
+			return true
+		}
+	}
+	return false
+}
+
+// andMatcher is a rbacMatcher that is successful if every child rbacMatcher matches.
+type andMatcher struct {
+	matchers []rbacMatcher
+}
+
+func (am *andMatcher) matches(args *EvaluateArgs) bool {
+	for _, matcher := range am.matchers {
+		if !matcher.matches(args) {
+			return false
+		}
+	}
+	return true
+}
+
+// alwaysMatcher is a rbacMatcher that will always match. This logically represents
+// an any rule for a permission or a principal.
+type alwaysMatcher struct {
+}
+
+func (am *alwaysMatcher) matches(args *EvaluateArgs) bool {
+	return true
+}
+
+// notMatcher is a rbacMatcher that nots an underlying rbacMatcher.
+type notMatcher struct {
+	matcherToNot rbacMatcher
+}
+
+func (nm *notMatcher) matches(args *EvaluateArgs) bool {
+	return !nm.matcherToNot.matches(args)
+}
+
+// headerMatcher is a rbacMatcher that matches on incoming HTTP Headers present in the incoming RPC.
+type headerMatcher struct {
+	headerMatcherInterface matcher.HeaderMatcherInterface
+}
+
+// TODO: Do you need errors on every node? Especially this one, where it doesn't
```

Comment:
If there are no conditions under which this can `error`, feel free to change the signature to not return an error value. This is anyways an internal package and we can change the API whenever we want if required.